flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/math/distributed/DistributedRowMatrix.scala

```diff
@@ -0,0 +1,179 @@
+import org.apache.flink.api.scala._
+import org.apache.flink.ml.math.Breeze._
+import org.apache.flink.ml.math.{Matrix => FlinkMatrix, _}
+
+/**
+  * Distributed row-major matrix representation.
+  * @param numRowsOpt If None, will be calculated from the DataSet.
+  * @param numColsOpt If None, will be calculated from the DataSet.
+  */
+class DistributedRowMatrix(data: DataSet[IndexedRow],
+                           numRowsOpt: Option[Int] = None,
+                           numColsOpt: Option[Int] = None)
+    extends DistributedMatrix {
+
+  lazy val getNumRows: Int = numRowsOpt match {
+    case Some(rows) => rows
+    case None => numRows.collect().head
+  }
+
+  lazy val getNumCols: Int = numColsOpt match {
+    case Some(cols) => cols
+    case None => numCols.collect().head
+  }
+
+  lazy val numRows: DataSet[Int] = numRowsOpt match {
+    case Some(rows) => data.getExecutionEnvironment.fromElements(rows)
+    case None => data.max("rowIndex").map(_.rowIndex + 1)
+  }
+
+  lazy val numCols: DataSet[Int] = numColsOpt match {
+    case Some(cols) => data.getExecutionEnvironment.fromElements(cols)
+    case None => data.first(1).map(_.values.size)
+  }
+
+  val getRowData = data
+
+  /**
+    * Collects the data in the form of a sequence of coordinates associated with their values.
+    * @return
+    */
+  def toCOO: Seq[(Int, Int, Double)] = {
+
+    val localRows = data.collect()
+
+    for (IndexedRow(rowIndex, vector) <- localRows;
+         (columnIndex, value) <- vector) yield (rowIndex, columnIndex, value)
+  }
+
+  /**
+    * Collects the data in the form of a SparseMatrix
+    * @return
+    */
+  def toLocalSparseMatrix: SparseMatrix = {
+    val localMatrix =
+      SparseMatrix.fromCOO(this.getNumRows, this.getNumCols, this.toCOO)
+    require(localMatrix.numRows == this.getNumRows)
+    require(localMatrix.numCols == this.getNumCols)
+    localMatrix
+  }
+
+  //TODO: convert to dense representation on the distributed matrix and collect it afterward
+  def toLocalDenseMatrix: DenseMatrix = this.toLocalSparseMatrix.toDenseMatrix
+
+  /**
+    * Apply a high-order function to couple of rows
+    * @param fun
+    * @param other
+    * @return
+    */
+  def byRowOperation(fun: (Vector, Vector) => Vector,
+                     other: DistributedRowMatrix): DistributedRowMatrix = {
+    val otherData = other.getRowData
+    require(this.getNumCols == other.getNumCols)
+    require(this.getNumRows == other.getNumRows)
+
+    val result = this.data
+      .fullOuterJoin(otherData)
+      .where("rowIndex")
+      .equalTo("rowIndex")(
+          (left: IndexedRow, right: IndexedRow) => {
+            val row1 = Option(left).getOrElse(IndexedRow(
+                    right.rowIndex,
+                    SparseVector.fromCOO(right.values.size, List((0, 0.0)))))
+            val row2 = Option(right).getOrElse(IndexedRow(
+                    left.rowIndex,
+                    SparseVector.fromCOO(left.values.size, List((0, 0.0)))))
```

Comment:
I would like to rewrite this block like following to avoid create unnecessary `IndexedRow` object:

```scala
val row1 = Option(left) match {
  case Some(row: IndexedRow) => row
  case None => IndexedRow(right.rowIndex, SparseVector.fromCOO(right.values.size, List((0, 0.0))))
}
val row2 = Option(right) match {
  case Some(row: IndexedRow) => row
  case None => IndexedRow(left.rowIndex, SparseVector.fromCOO(left.values.size, List((0, 0.0))))
}
```